server/s3_abort_multipart_action.cc

```diff
@@ -382,39 +484,76 @@ void S3AbortMultipartAction::startcleanup() {
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
 }
 
-void S3AbortMultipartAction::mark_oid_for_deletion() {
+void S3AbortMultipartAction::mark_oids_for_deletion() {
   s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
   assert(!oid_str.empty());
+  if (probable_del_rec_list.size() == 0) {
+    next();
+    return;
+  }
 
-  probable_delete_rec->set_force_delete(true);
+  std::map<std::string, std::string> oid_key_val_map;
+  for (auto& delete_rec : probable_del_rec_list) {
+    delete_rec->set_force_delete(true);
+    oid_key_val_map[delete_rec->get_key()] = delete_rec->to_json();
+  }
 
   if (!motr_kv_writer) {
     motr_kv_writer =
         mote_kv_writer_factory->create_motr_kvs_writer(request, s3_motr_api);
   }
   motr_kv_writer->put_keyval(global_probable_dead_object_list_index_oid,
-                             oid_str, probable_delete_rec->to_json(),
+                             oid_key_val_map,
                              std::bind(&S3AbortMultipartAction::next, this),
                              std::bind(&S3AbortMultipartAction::next, this));
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
 }
 
-void S3AbortMultipartAction::delete_object() {
+void S3AbortMultipartAction::delete_part() {
   s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
   // process to delete object
+  if (multipart_parts.size() == 0) {
+    next();
+    return;
+  }
   if (!motr_writer) {
     motr_writer = motr_writer_factory->create_motr_writer(
         request, object_multipart_metadata->get_oid());
   }
+
+  auto& part_metadata = multipart_parts.front();
+  struct m0_uint128 id = part_metadata->get_oid();
+  motr_writer->set_oid(id);
+  motr_writer->set_layout_id(part_metadata->get_layout_id());
   motr_writer->delete_object(
-      std::bind(&S3AbortMultipartAction::remove_probable_record, this),
+      std::bind(&S3AbortMultipartAction::delete_parts_successful, this),
       std::bind(&S3AbortMultipartAction::next, this),
-      object_multipart_metadata->get_layout_id());
+      part_metadata->get_layout_id());
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
 }
 
-void S3AbortMultipartAction::remove_probable_record() {
+void S3AbortMultipartAction::delete_parts_successful() {
   s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (multipart_parts.size() == 0) {
+    return;
+  }
+  s3_log(S3_LOG_INFO, request_id,
+         "Deleted old object oid "
+         "[%" SCNx64 " : %" SCNx64 "]",
+         (motr_writer->get_oid()).u_hi, (motr_writer->get_oid()).u_lo);
+  multipart_parts.erase(multipart_parts.begin());
```

Comment:
Erasing from beginning of vector would lead to acquiring new memory and copying elements. This would be significant, impact if number of elements in it is quite high (say, 10 thousand parts, which is AWS limit on number of parts)

Reply:
Make sense, will change that

Reply:
extracting from back and deleting from end